Update install guides for 1.4.0

tgraf opened the pull request:
This brings all install guides up to 1.4.0 including:

 * Dedicated guides for etcd-operator and external etcd
 * New EKS guide with eksctl and CoreDNS section based on new cilium-etcd-operator
 * New GKE guide using gcloud, auto-generated YAML and new DaemonSet for node initialization that works with auto-scaling
 * Re-restructure all getting started guides into sections
 * Consolidate all installation guides in one place

Related: https://github.com/cilium/cilium/projects/25

Fixes: #6783

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6812)
<!-- Reviewable:end -->

aanm commented on examples/kubernetes/node-init/node-init.yaml:
```diff
@@ -0,0 +1,74 @@
+      containers:
+        - name: node-init
+          image: gcr.io/google-containers/startup-script:v1
+          imagePullPolicy: Always
```
imagePullPolicy: IfNotPresent

aanm commented on Documentation/kubernetes/requirements.rst:
```diff
@@ -0,0 +1,95 @@
++---------------------+---------------------------------------------------------------+
+| Option              | Notes                                                         |
++---------------------+---------------------------------------------------------------+
+| 1.9                 |                                                               |
```
Missing 1.8

joestringer replied:
What is the point of this notes column? It seems to have no notes at all..?

tgraf replied:
I figured I would have notes to add as I complete the other sections but that never happend ;-). I've converted the table into a list.

ianvernon commented on Documentation/gettingstarted/k8s-install-etcd-operator.rst:
```diff
@@ -0,0 +1,98 @@
+
+This guides takes you through the steps required to set up Cilium on Kubernetes
+using the cilium-etcd-operator. The cilium-etcd-operator replaces the
+requirement for an external kvstore, you can learn more about it in the section
```
kvstore. You can learn more about it...

ianvernon commented on Documentation/gettingstarted/k8s-install-etcd-operator.rst:
```diff
@@ -0,0 +1,98 @@
+learn about the exact limitations of this deployment method.
+
+Should you encounter any issues during the installation, please refer to the
+:ref:`troubleshooting_k8s` section and / or seek help on `Slack channel`.
```
on the \`Slack channel\`

ianvernon commented on Documentation/gettingstarted/k8s-install-external-etcd.rst:
```diff
@@ -0,0 +1,121 @@
+
+This guide walks you through the steps required to set up Cilium on Kubernetes
+using an external etcd. Use of an external etcd provides better performance and
+is suitable for larger environment. If you are looking for a simple
```
environments

ianvernon commented on Documentation/gettingstarted/k8s-install-kops.rst:
```diff
@@ -62,23 +69,32 @@ Assuming you have all the prerequisites, run the following commands to create th
 
 
-``kops`` requires the creation of a dedicated s3 bucket in order to store the state and representation of the cluster. You will need to change the bucket name and provide your unique bucket name (for example a reverse of FQDN added with short description of the cluster). Also make sure to use the region where you will be deploying the cluster.
+``kops`` requires the creation of a dedicated s3 bucket in order to store the
```
s3 --> S3

ianvernon commented on Documentation/gettingstarted/k8s-install-kops.rst:
```diff
@@ -94,8 +110,12 @@ In this guide, we will use etcd version 3.1.11 and the latest CoreOS stable imag
+  creating the master and worker nodes. The number of master zones should be
+  * odd (1, 3, ...) for HA. For simplicity, you can just use 1 region.
+* The cluster ``NAME`` variable should end with ``k8s.local`` to use the gossip
+  protocol. If creating multiple clusters using the same kops user, then make
```
then make the

ianvernon commented on Documentation/gettingstarted/k8s-installers.rst:
```diff
@@ -0,0 +1,26 @@
+
+The following list includes the Kubernetes installer integrations we are aware
+of. If your preferred installer is not in the list, you can always fall back to
+the standard :ref:`k8s_install_etcd_operator` guide which works independent of
```
independently

ianvernon commented on Documentation/kubernetes/requirements.rst:
```diff
@@ -0,0 +1,95 @@
+Enable automatic node CIDR allocation (Recommended)
+===================================================
+
+Kubernetes has the capability to automatically allocate and assign per node IP
```
assign a per-node

ianvernon reviewed (commented):
Minor nits. Rest LGTM

tgraf committed:
kubernetes: Provide a node-init DaemonSet for GKE

This enables auto-scaling and automatic enablement of nodes without requiring
to execute additional commands. It can be further improved with tolerations to
avoid scheduling before the node-init DaemonSet has run. It is also potentially
useful to EKS.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
kubernetes: Generate all-in-one yaml for node init

Diff to github.com/cilium/gke/cilium-deployment.yaml for easier review:
```
--- cilium-deployment.yaml	2019-01-28 19:56:35.000000000 -0800
+++ 1.12/cilium-with-node-init.yaml	2019-01-28 20:07:02.000000000 -0800
@@ -131,6 +131,12 @@
   # a container. When the Cilium DaemonSet is removed, the BPF programs will
   # be kept in the interfaces unless this option is set to "true".
   flannel-uninstall-on-exit: "false"
+  # Installs a BPF program to allow for policy enforcement in already running
+  # containers managed by Flannel.
+  # NOTE: This requires Cilium DaemonSet to be running in the hostPID.
+  # To run in this mode in Kubernetes change the value of the hostPID from
+  # false to true. Can be found under the path `spec.spec.hostPID`
+  flannel-manage-existing-containers: "false"
 ---
 apiVersion: apps/v1
 kind: DaemonSet
@@ -271,8 +277,14 @@
               key: flannel-uninstall-on-exit
               name: cilium-config
               optional: true
-        image: docker.io/cilium/cilium:v1.4.0-rc4
-        imagePullPolicy: IfNotPresent
+        - name: CILIUM_FLANNEL_MANAGE_EXISTING_CONTAINERS
+          valueFrom:
+            configMapKeyRef:
+              key: flannel-manage-existing-containers
+              name: cilium-config
+              optional: true
+        image: docker.io/cilium/cilium:latest
+        imagePullPolicy: Always
         lifecycle:
           postStart:
             exec:
@@ -339,6 +351,7 @@
           readOnly: true
       dnsPolicy: ClusterFirstWithHostNet
       hostNetwork: true
+      hostPID: false
       initContainers:
       - command:
         - /init-container.sh
@@ -505,8 +518,8 @@
               key: AWS_DEFAULT_REGION
               name: cilium-aws
               optional: true
-        image: docker.io/cilium/operator:v1.4.0-rc4
-        imagePullPolicy: IfNotPresent
+        image: docker.io/cilium/operator:latest
+        imagePullPolicy: Always
         name: cilium-operator
         volumeMounts:
         - mountPath: /var/lib/etcd-config
@@ -534,9 +547,10 @@
           defaultMode: 420
           optional: true
           secretName: cilium-etcd-secrets
+
 ---
-apiVersion: v1
 kind: ServiceAccount
+apiVersion: v1
 metadata:
   name: cilium-operator
   namespace: cilium
@@ -545,6 +559,7 @@
 kind: ClusterRole
 metadata:
   name: cilium-operator
+  namespace: cilium
 rules:
 - apiGroups:
   - ""
@@ -591,6 +606,7 @@
 kind: ClusterRole
 metadata:
   name: cilium-etcd-operator
+  namespace: cilium
 rules:
 - apiGroups:
   - etcd.database.coreos.com
@@ -660,6 +676,7 @@
 kind: ClusterRoleBinding
 metadata:
   name: cilium-etcd-operator
+  namespace: cilium
 roleRef:
   apiGroup: rbac.authorization.k8s.io
   kind: ClusterRole
@@ -673,6 +690,7 @@
 kind: ClusterRole
 metadata:
   name: etcd-operator
+  namespace: cilium
 rules:
 - apiGroups:
   - etcd.database.coreos.com
@@ -726,6 +744,7 @@
 kind: ClusterRoleBinding
 metadata:
   name: etcd-operator
+  namespace: cilium
 roleRef:
   apiGroup: rbac.authorization.k8s.io
   kind: ClusterRole
@@ -735,8 +754,8 @@
   name: cilium-etcd-sa
   namespace: cilium
 ---
-apiVersion: v1
 kind: ServiceAccount
+apiVersion: v1
 metadata:
   name: cilium-etcd-operator
   namespace: cilium
```

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Restructure getting started guides

This completely restructures the installation relevant getting started guides
and puts them all in one place. It also provides detailed instructions for
the etcd-operator and provides a separate guide when using an external etcd.

* New EKS guide using eksctl, relies on #6805
* GKE guide

Fixes: #6783

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
kubernetes: Provide all-on-one YAML for external etcd

Signed-off-by: Thomas Graf <thomas@cilium.io>
